src/HasByNonDependentSubqueryMacro.php

```diff
@@ -146,10 +147,40 @@ protected function adjustArgumentTypeOfOptionalConstraints(callable $constraint,
     {
         $reflection = ReflectionCallable::from($constraint);
 
-        return $reflection->getNumberOfParameters() > 0
+        $type = $reflection->getNumberOfParameters() > 0
             && ($parameter = $reflection->getParameters()[0])->hasType()
-            && \is_a($parameter->getType()->getName(), Builder::class, true)
-            ? $relation->getQuery()
-            : $relation;
+            ? $parameter->getType()
+            : null;
+
+        if (!$type) {
+            return $relation;
+        }
+
+        return $this->determineArgumentType($type, $relation);
+    }
+
+    /**
+     * @param  \ReflectionType                                                                        $type
+     * @param  \Illuminate\Database\Eloquent\Relations\Relation                                       $relation
+     * @return \Illuminate\Database\Eloquent\Builder|\Illuminate\Database\Eloquent\Relations\Relation
+     */
+    protected function determineArgumentType(ReflectionType $type, Relation $relation)
+    {
+        // https://www.php.net/manual/en/class.reflectionnamedtype.php
+        if (method_exists($type, 'getName')) {
+            return \is_a($type->getName(), Builder::class, true)
+                ? $relation->getQuery()
+                : $relation;
+        }
+
+        // https://www.php.net/manual/en/class.reflectionuniontype.php
+        if (method_exists($type, 'getTypes')) {
+            /** @var ReflectionType $firstType */
+            $firstType = $type->getTypes()[0];
```

Comment:
Also, it would be great if adjusting by return type of getTypes() could be made **in any order!**

Reply:
Loop through them and the first one that is a `Builder`, return a query for that one?